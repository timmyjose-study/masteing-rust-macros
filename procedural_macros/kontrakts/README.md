Working through (and hacking on ) the [Ferrous Systems procedural macros example](https://github.com/ferrous-systems/testing-proc-macros) for the purpose of mastering procedural macros in Rust.