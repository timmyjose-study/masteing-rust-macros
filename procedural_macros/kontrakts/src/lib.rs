use proc_macro::TokenStream;
use proc_macro_error::proc_macro_error;

//mod analyzer;
//mod codegen;
//mod ir;
mod parser;

use parser::{parse, Ast};
//use analyzer::{Model, analyze};
//use ir::{Ir, lower};
//use codegen::gen_code;

#[proc_macro_attribute]
#[proc_macro_error]
pub fn contracts(args: TokenStream, item: TokenStream) -> TokenStream {
    let ast = parse(args.into(), item.into());
    //let model = analyze(ast);
    //let ir = lower(model);
    //let rust_tok_stream = gen_code(ir);

    //rust_tok_stream
    todo!()
}
