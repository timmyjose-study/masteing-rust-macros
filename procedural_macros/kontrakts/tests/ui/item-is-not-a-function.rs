use kontrakts::contracts;

#[contracts]
struct S;

fn main() {}