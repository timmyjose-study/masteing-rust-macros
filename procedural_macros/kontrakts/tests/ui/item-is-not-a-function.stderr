error: item is not a function

         = help: `#[contracts]` can only be used on functions

 --> tests/ui/item-is-not-a-function.rs:4:1
  |
4 | struct S;
  | ^^^^^^^^^
